packages/markdown/src/index.ts

```diff
@@ -64,35 +71,68 @@ export function createProcessor({
+  allowHtml = true,
+  interpretMath = true,
 }: {
   mdastVisitors?: ((ast: MdastRoot) => undefined)[];
   hastVisitors?: ((ast: HastRoot) => undefined)[];
   sanitize?: boolean;
+  allowHtml?: boolean;
+  interpretMath?: boolean;
 } = {}) {
-  const htmlConversion = (mdastVisitors ?? [])
-    .reduce<Processor<MdastRoot, MdastRoot, MdastRoot | undefined>>(
-      (processor, visitor) => processor.use(makeHandler(visitor)),
-      unified().use(remarkParse).use(remarkMath),
-    )
-    .use(makeHandler(visitMathBlock))
-    .use(remarkGfm)
-    .use(remark2rehype, { allowDangerousHtml: true })
-    .use(rehypeRaw);
-  return (hastVisitors ?? [])
-    .reduce(
-      (processor, visitor) => processor.use(makeHandler(visitor)),
-      sanitize ? htmlConversion.use(rehypeSanitize) : htmlConversion,
-    )
-    .use(rehypeStringify);
+  const plugins: (Plugin<any, any, any> | PluginTuple<any, any, any>)[] = [
+    remarkParse,
+    ...(interpretMath ? [remarkMath] : []),
+    ...(mdastVisitors ?? []).map((visitor) => makeHandler(visitor)),
+    ...(interpretMath ? [makeHandler(visitMathBlock)] : []),
+    remarkGfm,
+    [remark2rehype, { allowDangerousHtml: allowHtml }],
+    ...(!allowHtml ? [] : [rehypeRaw]),
+    ...(sanitize ? [rehypeSanitize] : []),
+    ...(hastVisitors ?? []).map((visitor) => makeHandler(visitor)),
+    rehypeStringify,
+  ];
+
+  return plugins.reduce((processor: Processor, plugin) => {
+    if (Array.isArray(plugin)) return processor.use(...plugin);
+    return processor.use(plugin);
+  }, unified());
 }
 
-const defaultProcessor = createProcessor();
-const inlineProcessor = createProcessor({ hastVisitors: [visitCheckSingleParagraph] });
+const processorCache = new Map<string, Processor>();
+function getProcessor(options: {
+  inline: boolean;
+  allowHtml: boolean;
+  interpretMath: boolean;
+}): Processor {
+  const key = `${options.inline}:${options.allowHtml}:${options.interpretMath}`;
+  let processor = processorCache.get(key);
+  if (!processor) {
+    processor = createProcessor({
+      hastVisitors: options.inline ? [visitCheckSingleParagraph] : [],
+      allowHtml: options.allowHtml,
+      interpretMath: options.interpretMath,
+    });
+    processorCache.set(key, processor);
+  }
+  return processor;
+}
 
 /**
  * Converts markdown to HTML. If `inline` is true, and the result fits a single
  * paragraph, the content is returned inline without the paragraph tag.
  */
-export async function markdownToHtml(original: string, { inline }: { inline?: boolean } = {}) {
-  return (await (inline ? inlineProcessor : defaultProcessor).process(original)).value.toString();
+export async function markdownToHtml(
+  original: string,
+  {
+    inline = false,
+    allowHtml = true,
+    interpretMath = true,
+  }: { inline?: boolean; allowHtml?: boolean; interpretMath?: boolean } = {},
+) {
+  return (
+    await getProcessor({ inline, allowHtml, interpretMath }).process(original)
+  ).value.toString();
```

Comment:
Tiny nit, IMO splitting it up makes it slightly easier to read.

```suggestion
) {
  const processor = getProcessor({ inline, allowHtml, interpretMath });
  return (await processor.process(original)).value.toString();
```